File: python/LinkedList.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
        
    def print_all(self):
        current_node = self.head
        result = "["
        while current_node:
            result += str(current_node.data)
            current_node = current_node.next
            
            if current_node:
                result += ", "
        result += "]"
        print(result)
    
    def clear(self):
        self.head = None
        self.size = 0
        
    def insert_at(self, index, data):
        if index < 0 or index > self.size:
            raise IndexError("추가할 수 있는 범위를 넘어갔습니다.")
        
        new_node = Node(data)
        
        if index == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            current_node = self.head
            for _ in range(index - 1):
                current_node = current_node.next
            new_node.next = current_node.next
            current_node.next = new_node
        
        self.size += 1
        
    def insert_last(self, data):
        self.insert_at(self.size, data)
        
    def delete_at(self, index):
        if index < 0 or index > self.size:
            raise IndexError("삭제할 수 있는 범위를 넘어갔습니다.")
        
        if index == 0:
            deleted_node = self.head
            self.head = self.head.next
        else:
            current_node = self.head
            for _ in range(index - 1):
                current_node = current_node.next
            deleted_node = current_node.next
            current_node.next = current_node.next.next
        
        self.size -= 1
        return deleted_node
        
    def delete_last(self):
        self.delete_at(self.size - 1)
        
    def get_node_at(self, index):
        if index < 0 or index >= self.size:
            raise IndexError("범위를 넘어갔습니다.")
        
        current_node = self.head
        for _ in range(index):
            current_node = current_node.next
        return current_node

File: js/LinkedList.mjs
class Node {
    constructor(data, next = null) {
        this.data = data;
        this.next = next;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.count = 0;
    }

    insertAt(index, data) {
        if (index < 0 || index > this.count) {
            throw new Error('범위를 넘어갔습니다.');
        }

        let newNode = new Node(data);

        if (index === 0) {
            newNode.next = this.head;
            this.head = newNode;
        } else {
            let current = this.head;

            for (let i = 0; i < index - 1; i++) {
                current = current.next;
            }

            newNode.next = current.next;
            current.next = newNode;
        }
    }

    printAll() {
        let current = this.head;
        let result = '[';

        while (current) {
            result += current.data;
            current = current.next;

            if (current != null) {
                result += ', ';
            }
        }

        result += ']';
        console.log(result);
    }

    clear() {
        this.head = null;
        this.count = 0;
    }

    insertLast(data) {
        this.insertAt(this.count, data);
    }

    deleteAt(index) {
        if (index < 0 || index > this.count) {
            throw new Error('제거할 수 없습니다.');
        }

        let current = this.head;

        if (index === 0) {
            let deletedNode = current;
            this.head = current.next;
            this.count--;
            return deletedNode;
        } else {
            for (let i = 0; i < index - 1; i++) {
                current = current.next;
            }

            let deletedNode = current.next;
            current.next = current.next.next;
            this.count--;
            return deletedNode;
        }
    }

    deleteLast() {
        this.deleteAt(this.count - 1);
    }

    getNodeAt(index) {
        if (index < 0 || index > this.count) {
            throw new Error('범위를 넘어갔습니다.');
        }

        let current = this.head;

        for (let i = 0; i < index; i++) {
            current = current.next;
        }
        return current;
    }
}

export { Node, LinkedList }